### PR TITLE
Add File Type Validation To Store Logo

### PR DESCRIPTION
**Pull request by MatthewKennedy**
- Adds active_storage_validations gem.
- Validates image file type. (.png, jpg, jpeg, gif, svg)
- Adds some CSS to stop an svg blowing up huge on the edit store page.

### CODE REVIEW

File: backend/app/controllers/spree/admin/stores_controller.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class StoresController < Spree::Admin::BaseController
       before_action :set_default_locale, only: :new
       before_action :normalize_supported_currencies, only: [:create, :update]
       before_action :set_default_country_id, only: :new
-      before_action :load_all_countries, only: [:new, :edit, :update]
+      before_action :load_all_countries, only: [:new, :edit, :update, :create]
```
Comment (damianlegawiec) [inline backend/app/controllers/spree/admin/stores_controller.rb:9]: @MatthewKennedy can you extract this change to a separate PR? thanks!

File: backend/spree_backend.gemspec [hunk 2]
```diff
@@ -28,11 +28,12 @@ Gem::Specification.new do |s|
...
-  s.add_dependency 'jquery-ui-rails', '~> 6.0.1'
-  s.add_dependency 'select2-rails',   '~> 3.5.0'
+  s.add_dependency 'active_storage_validations', '~> 0.8.0'
+  s.add_dependency 'bootstrap',                  '>= 4.3.1', '< 4.6.0'
+  s.add_dependency 'glyphicons',                 '~> 1.0.2'
+  s.add_dependency 'flatpickr',                  '~> 4.6.0'
+  s.add_dependency 'inline_svg',                 '~> 1.5'
+  s.add_dependency 'jquery-rails',               '~> 4.3'
+  s.add_dependency 'jquery-ui-rails',            '~> 6.0.1'
+  s.add_dependency 'select2-rails',              '~> 3.5.0'
```
Comment (damianlegawiec) [inline backend/spree_backend.gemspec:38]: @MatthewKennedy let's only modify one line here, so we're not losing track of previous changes

File: core/app/models/spree/store.rb [hunk 3]
```diff
@@ -17,6 +17,8 @@ class Store < Spree::Base
 
     has_one_attached :logo
 
+    validates :logo, content_type: [:png, :jpg, :jpeg, :gif]
```
Comment (damianlegawiec) [inline core/app/models/spree/store.rb:20]: @MatthewKennedy what about SVG?

File: backend/app/views/spree/admin/stores/_form.html.erb [hunk 4]
```diff
@@ -9,7 +9,7 @@
   <div class="card-body">
     <%= f.field_container :logo, class: ['form-group'] do %>
       <% if @store.logo.attached? %>
-        <%= image_tag main_app.url_for(@store.logo.variant(resize: '244x104>')) %>
```
Comment (damianlegawiec) [inline backend/app/views/spree/admin/stores/_form.html.erb:12]: @MatthewKennedy hmm this resize here was a safeguard, if someone would upload a huge file

File: backend/spree_backend.gemspec [hunk 5]
```diff
@@ -28,11 +28,12 @@ Gem::Specification.new do |s|
...
   s.add_dependency 'spree_core', s.version
 
-  s.add_dependency 'bootstrap',       '>= 4.3.1', '< 4.6.0'
-  s.add_dependency 'glyphicons',      '~> 1.0.2'
-  s.add_dependency 'flatpickr',       '~> 4.6.0'
-  s.add_dependency 'inline_svg',      '~> 1.5'
-  s.add_dependency 'jquery-rails',    '~> 4.3'
-  s.add_dependency 'jquery-ui-rails', '~> 6.0.1'
-  s.add_dependency 'select2-rails',   '~> 3.5.0'
+  s.add_dependency 'active_storage_validations', '~> 0.8.0'
```
Comment (damianlegawiec) [inline backend/spree_backend.gemspec:31]:
```suggestion
  s.add_dependency 'active_storage_validations', '~> 0.8'
```

File: backend/lib/spree/backend.rb [hunk 6]
```diff
@@ -7,6 +7,7 @@
 require 'glyphicons'
 require 'popper_js'
 require 'inline_svg'
+require 'active_storage_validations'
```
Comment (damianlegawiec) [inline backend/lib/spree/backend.rb:10]: @MatthewKennedy, for now, we can add this gem, in Rails 6.1 this will be part of the Rails core

File: backend/spree_backend.gemspec (hunk 2 above)
Comment (MatthewKennedy) [reply backend/spree_backend.gemspec:38]: Can do.

File: backend/app/views/spree/admin/stores/_form.html.erb (hunk 4 above)
Comment (MatthewKennedy) [reply backend/app/views/spree/admin/stores/_form.html.erb:12]: With an svg the resize will cause an error, But you can validate file size with eh active_storage_validations gem?

File: core/app/models/spree/store.rb (hunk 3 above)
Comment (MatthewKennedy) [reply core/app/models/spree/store.rb:20]: I think I tried .svg as a file type, but the resize causes an error front end and back.

File: backend/app/controllers/spree/admin/stores_controller.rb (hunk 1 above)
Comment (MatthewKennedy) [reply backend/app/controllers/spree/admin/stores_controller.rb:9]: Sure thats not a problem.

File: core/app/models/spree/store.rb (hunk 3 above)
Comment (MatthewKennedy) [reply core/app/models/spree/store.rb:20]: @damianlegawiec https://github.com/spree/spree/blob/caeb26d6761851e7b032832f48e1bee39418094c/core/app/helpers/spree/base_helper.rb#L44 

The resize in this method causes an error on svg file types on the front end. I'll just fire up this branch and test that to make sure.

Comment (damianlegawiec) [reply core/app/models/spree/store.rb:20]: @MatthewKennedy you're correct - we should only use resize if the file is an image

Comment (MatthewKennedy) [reply core/app/models/spree/store.rb:20]: @damianlegawiec if you take off the resize on the backend, and then upload an .svg  it won't display, I suspect this may be because it is used within img tags, or maybe it's something deeper than that, I haven't looked just yet.

Screenshot 2020-08-07 at 13 56 39

I suspect that using an .svg as store logo would need to be handled completely differently to you're regular images.

File: backend/app/assets/stylesheets/spree/backend/views/_configurations_stores.scss [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#store_logo_field {
+    max-width: 200px;
+    img{
+        max-width: 100%;
+    }
+}
```
Comment (MohamedBarakats) [inline backend/app/assets/stylesheets/spree/backend/views/_configurations_stores.scss:6]: end of line

Comment (MohamedBarakats) [inline backend/app/assets/stylesheets/spree/backend/views/_configurations_stores.scss:6]: new line at end of file

**Review by mbarakats — CHANGES_REQUESTED** [review]
